File: tests/test_market_validation.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Tests for market_validation.py — the Internal Market Validation System
(Founder Directive, 2026-08-14). Pure stdlib unittest, same isolation
discipline as tests/test_market_evidence.py (temp JSONL path, never
touches the real data/ ledger).

    python -m unittest tests.test_market_validation -v
"""

import json
import os
import sys
import tempfile
import unittest

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
import market_validation  # noqa: E402


def _valid_payload(**overrides):
    payload = {
        "q1_frequency": "weekly",
        "q2_intent": "maybe",
        "q3_pain": "I waste hours checking whether citations are still good law.",
        "source": "linkedin",
        "professional_role": "Solo attorney",
        "practice_area": "Family law",
        "contact": "lawyer@example.com",
        "session_id": "sess-test-1",
    }
    payload.update(overrides)
    return payload


class ValidationStorageTests(unittest.TestCase):

    def setUp(self):
        self._tmp = tempfile.NamedTemporaryFile(prefix="val_resp_", suffix=".jsonl", delete=False)
        self._tmp.close()
        self._path = self._tmp.name
        self._old_env = os.environ.get("VALIDATION_RESPONSES_PATH")
        os.environ["VALIDATION_RESPONSES_PATH"] = self._path

    def tearDown(self):
        if self._old_env is None:
            os.environ.pop("VALIDATION_RESPONSES_PATH", None)
        else:
            os.environ["VALIDATION_RESPONSES_PATH"] = self._old_env
        if os.path.exists(self._path):
            os.remove(self._path)

    def test_record_appends_one_jsonl_row_and_returns_timestamp(self):
        row = market_validation.record_response(_valid_payload())
        self.assertIn("timestamp", row)
        with open(self._path, "r", encoding="utf-8") as f:
            lines = [l for l in f if l.strip()]
        self.assertEqual(len(lines), 1)
        stored = json.loads(lines[0])
        self.assertEqual(stored["q1_frequency"], "weekly")
        self.assertEqual(stored["source"], "linkedin")

    def test_validate_rejects_bad_q1(self):
        with self.assertRaises(ValueError) as ctx:
            market_validation.record_response(_valid_payload(q1_frequency="always"))
        self.assertIn("q1_frequency", str(ctx.exception))

    def test_validate_rejects_bad_q2(self):
        with self.assertRaises(ValueError) as ctx:
            market_validation.record_response(_valid_payload(q2_intent="definitely"))
        self.assertIn("q2_intent", str(ctx.exception))

    def test_validate_rejects_empty_q3(self):
        with self.assertRaises(ValueError) as ctx:
            market_validation.record_response(_valid_payload(q3_pain="  "))
        self.assertIn("q3_pain", str(ctx.exception))

    def test_validate_rejects_non_dict_payload(self):
        with self.assertRaises(ValueError):
            market_validation.record_response("not an object")

    def test_unknown_source_normalized_to_other(self):
        row = market_validation.record_response(_valid_payload(source="telegram"))
        self.assertEqual(row["source"], "other")

    def test_duplicate_is_rejected_not_doubled(self):
        market_validation.record_response(_valid_payload())
        with self.assertRaises(ValueError) as ctx:
            market_validation.record_response(_valid_payload())
        self.assertIn("duplicate", str(ctx.exception))
        with open(self._path, "r", encoding="utf-8") as f:
            lines = [l for l in f if l.strip()]
        self.assertEqual(len(lines), 1)

    def test_optional_fields_default_to_empty(self):
        row = market_validation.record_response(_valid_payload(professional_role=None, practice_area="", contact=None, session_id=""))
        self.assertEqual(row["professional_role"], "")
        self.assertEqual(row["practice_area"], "")
        self.assertEqual(row["contact"], "")
        self.assertEqual(row["session_id"], "")

    def test_sample_request_and_waitlist_stored(self):
        row = market_validation.record_response(_valid_payload(sample_request="yes", waitlist="yes"))
        self.assertEqual(row["sample_request"], "yes")
        self.assertEqual(row["waitlist"], "yes")

    def test_rejects_invalid_sample_request(self):
        with self.assertRaises(ValueError) as ctx:
            market_validation.record_response(_valid_payload(sample_request="maybe"))
        self.assertIn("sample_request", str(ctx.exception))

    def test_rejects_invalid_waitlist(self):
        with self.assertRaises(ValueError) as ctx:
            market_validation.record_response(_valid_payload(waitlist="certainly"))
        self.assertIn("waitlist", str(ctx.exception))


class ValidationAggregationTests(unittest.TestCase):

    def setUp(self):
        self._tmp = tempfile.NamedTemporaryFile(prefix="val_resp_", suffix=".jsonl", delete=False)
        self._tmp.close()
        self._path = self._tmp.name
        self._old_env = os.environ.get("VALIDATION_RESPONSES_PATH")
        os.environ["VALIDATION_RESPONSES_PATH"] = self._path
        market_validation.record_response(_valid_payload(
            session_id="a", q1_frequency="several_times_per_week", q2_intent="yes", source="linkedin",
            sample_request="yes", waitlist="yes"))
        market_validation.record_response(_valid_payload(
            session_id="b", q1_frequency="weekly", q2_intent="maybe", source="facebook",
            waitlist="yes"))
        market_validation.record_response(_valid_payload(
            session_id="c", q1_frequency="rarely", q2_intent="no", source="direct",
            professional_role="", practice_area=""))

    def tearDown(self):
        if self._old_env is None:
            os.environ.pop("VALIDATION_RESPONSES_PATH", None)
        else:
            os.environ["VALIDATION_RESPONSES_PATH"] = self._old_env
        if os.path.exists(self._path):
            os.remove(self._path)

    def test_aggregate_counts_total_and_qualified(self):
        s = market_validation.aggregate()
        self.assertEqual(s["total_responses"], 3)
        self.assertEqual(s["qualified_responses"], 2)
        self.assertEqual(s["pain_signals"], 3)
        self.assertEqual(s["weekly_or_more_pain"], 2)

    def test_aggregate_intent_counts(self):
        s = market_validation.aggregate()
        self.assertEqual(s["q2_yes"], 1)
        self.assertEqual(s["q2_maybe"], 1)
        self.assertEqual(s["q2_no"], 1)

    def test_aggregate_conversion_rate_is_qualified_yes(self):
        s = market_validation.aggregate()
        self.assertEqual(s["qualified_conversion_rate_pct"], 50.0)

    def test_aggregate_source_breakdown(self):
        s = market_validation.aggregate()
        self.assertEqual(s["source_breakdown"], {"linkedin": 1, "facebook": 1, "direct": 1})

    def test_aggregate_sample_requests_and_waitlist(self):
        s = market_validation.aggregate()
        self.assertEqual(s["sample_requests"], 1)
        self.assertEqual(s["waitlist_signups"], 2)

    def test_aggregate_visits_counts_validation_page_only(self):
        self._views_tmp = tempfile.NamedTemporaryFile(prefix="val_views_", suffix=".jsonl", delete=False)
        self._views_tmp.close()
        self._views_path = self._views_tmp.name
        market_validation.record_visit("linkedin", page_views_path=self._views_path)
        market_validation.record_visit("facebook", page_views_path=self._views_path)
        s = market_validation.aggregate(page_views_path=self._views_path)
        self.assertEqual(s["visits"], 2)
        views = market_validation.read_visits(self._views_path)
        self.assertEqual(len(views), 2)
        self.assertTrue(all(v["page_id"] == market_validation.VALIDATION_PAGE_ID for v in views))
        if os.path.exists(self._views_path):
            os.remove(self._views_path)

    def test_aggregate_never_exposes_pii_or_raw_text(self):
        s = market_validation.aggregate()
        dumped = json.dumps(s)
        self.assertNotIn("lawyer@example.com", dumped)
        self.assertNotIn("I waste hours", dumped)
        self.assertNotIn("contact", dumped)

    def test_aggregate_returns_top_themes_from_real_text(self):
        s = market_validation.aggregate()
        self.assertTrue(any(t["theme"] == "citations" for t in s["top_pain_themes"]))
        self.assertLessEqual(len(s["top_pain_themes"]), 10)

    def test_aggregate_empty_ledger_is_safe(self):
        os.remove(self._path)
        s = market_validation.aggregate()
        self.assertEqual(s["total_responses"], 0)
        self.assertIsNone(s["qualified_conversion_rate_pct"])
        self.assertEqual(s["source_breakdown"], {})
        self.assertEqual(s["top_pain_themes"], [])

    def test_corrupt_line_does_not_break_aggregate(self):
        with open(self._path, "a", encoding="utf-8") as f:
            f.write("{ this is not json\n")
        s = market_validation.aggregate()
        self.assertEqual(s["total_responses"], 3)


if __name__ == "__main__":
    unittest.main()

File: tests/test_market_validation.js
// Tests for the Internal Market Validation System (Founder Directive,
// 2026-08-14): public validation page + public submit, auth-gated
// internal dashboard, and the no-PII protection rule. Boots the real
// server.js (same pattern as tests/test_customer_front_door.js), with
// VALIDATION_RESPONSES_PATH pointed at a temp ledger so these tests
// never write into the real data/validation_responses.jsonl file.
//
//   node --test tests/test_market_validation.js

const test = require('node:test');
const assert = require('node:assert/strict');
const fs = require('fs');
const os = require('os');
const path = require('path');
const { spawn } = require('child_process');

const PORT = 3299; // distinct from test_customer_front_door.js's 3298 and test_api_contract.js's 3199
const BASE_URL = `http://localhost:${PORT}`;
const TEST_PASSWORD = 'validation-test-password';
const REPO_ROOT = path.join(__dirname, '..');
const TEMP_LEDGER = path.join(os.tmpdir(), `val_ledger_${process.pid}.jsonl`);

let serverProcess;

async function waitForServer(timeoutMs = 15000) {
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    try {
      const res = await fetch(`${BASE_URL}/api/dashboard`);
      if (res.status) return;
    } catch { /* not up yet */ }
    await new Promise((r) => setTimeout(r, 200));
  }
  throw new Error('server did not become ready in time');
}

function login() {
  return fetch(`${BASE_URL}/api/mission-control/login`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ password: TEST_PASSWORD }),
  });
}

const VALID_PAYLOAD = {
  q1_frequency: 'weekly',
  q2_intent: 'maybe',
  q3_pain: 'I lose hours verifying whether my citations are still good law.',
  source: 'linkedin',
  professional_role: 'Solo attorney',
  practice_area: 'Family law',
  contact: 'real.prospect@example.com',
};

test.before(async () => {
  if (fs.existsSync(TEMP_LEDGER)) fs.unlinkSync(TEMP_LEDGER);
  serverProcess = spawn(process.execPath, ['server.js'], {
    cwd: REPO_ROOT,
    env: { ...process.env, PORT: String(PORT), MISSION_CONTROL_PASSWORD: TEST_PASSWORD, VALIDATION_RESPONSES_PATH: TEMP_LEDGER },
  });
  await waitForServer();
});

test.after(() => {
  if (serverProcess) serverProcess.kill();
  if (fs.existsSync(TEMP_LEDGER)) fs.unlinkSync(TEMP_LEDGER);
});

test('public validation page is reachable unauthenticated and is a real responsive HTML page', async () => {
  const res = await fetch(`${BASE_URL}/market-validation.html`);
  assert.equal(res.status, 200);
  const body = await res.text();
  assert.match(body, /Professional Legal Research — Help Us Validate This/, 'must carry the exact founder headline');
  assert.match(body, /\$194/, 'must state the $194 future price');
  assert.match(body, /NOT a launched product/, 'must declare validation-status, not a live product');
  assert.match(body, /name="viewport" content="width=device-width/, 'must be mobile responsive');
  assert.match(body, /@media/, 'must have a real responsive breakpoint');
});

test('internal dashboard PAGE is protected: unauthenticated -> redirected to login', async () => {
  // requireMissionControlAuth sends non-API HTML GETs to the login page
  // (302), not a 401 JSON — the established pattern every other protected
  // Mission Control page follows. The point is the page content must never
  // be served unauthenticated.
  const res = await fetch(`${BASE_URL}/validation-dashboard.html`, { redirect: 'manual' });
  assert.equal(res.status, 302);
  assert.ok(res.headers.get('location').includes('mission_control_login.html'), 'must redirect to the real login page');
});

test('internal dashboard API is protected: unauthenticated -> 401', async () => {
  const res = await fetch(`${BASE_URL}/api/validation/dashboard`);
  assert.equal(res.status, 401);
});

test('public submit records exactly one real response', async () => {
  const res = await fetch(`${BASE_URL}/api/validation/submit`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(VALID_PAYLOAD),
  });
  assert.equal(res.status, 200);
  const body = await res.json();
  assert.equal(body.success, true);
  assert.equal(body.recorded, true);
  const lines = fs.readFileSync(TEMP_LEDGER, 'utf8').split('\n').filter(Boolean);
  assert.equal(lines.length, 1);
});

test('public submit rejects a bad source at the boundary', async () => {
  const res = await fetch(`${BASE_URL}/api/validation/submit`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ...VALID_PAYLOAD, source: 'tiktok' }),
  });
  assert.equal(res.status, 400);
  const body = await res.json();
  assert.equal(body.success, false);
});

test('public submit rejects invalid q2 at the boundary', async () => {
  const res = await fetch(`${BASE_URL}/api/validation/submit`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ...VALID_PAYLOAD, q2_intent: 'definitely' }),
  });
  assert.equal(res.status, 400);
  const body = await res.json();
  assert.equal(body.success, false);
});

test('public submit rejects a duplicate (no phantom second response)', async () => {
  const res = await fetch(`${BASE_URL}/api/validation/submit`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(VALID_PAYLOAD),
  });
  assert.equal(res.status, 400);
  const body = await res.json();
  assert.equal(body.success, false);
  const lines = fs.readFileSync(TEMP_LEDGER, 'utf8').split('\n').filter(Boolean);
  assert.equal(lines.length, 1);
});

test('dashboard API is reachable with a real login and exposes only aggregates', async () => {
  const loginRes = await login();
  assert.equal(loginRes.status, 200);
  const cookie = loginRes.headers.get('set-cookie').split(';')[0];

  const res = await fetch(`${BASE_URL}/api/validation/dashboard`, { headers: { Cookie: cookie } });
  assert.equal(res.status, 200);
  const body = await res.json();
  const s = body.summary;
  assert.equal(s.total_responses, 1);
  assert.equal(s.qualified_responses, 1);
  assert.equal(s.q2_maybe, 1);
  assert.equal(s.source_breakdown.linkedin, 1);
  assert.ok(Array.isArray(s.top_pain_themes));
  const dumped = JSON.stringify(body);
  assert.ok(!dumped.includes('real.prospect@example.com'), 'aggregate must never expose a contact email');
  assert.ok(!dumped.includes('I lose hours'), 'aggregate must never expose verbatim q3 text');
});

test('dashboard page is reachable with a real login', async () => {
  const loginRes = await login();
  assert.equal(loginRes.status, 200);
  const cookie = loginRes.headers.get('set-cookie').split(';')[0];
  const res = await fetch(`${BASE_URL}/validation-dashboard.html`, { headers: { Cookie: cookie } });
  assert.equal(res.status, 200);
  const body = await res.text();
  assert.match(body, /<title>Market Validation Dashboard/, 'must be the internal dashboard page');
});
